### ISSUE TITLE
[JENKINS-57542] git checkout over ssh fails on windows agents started from msys2 - missing bat file

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-client-plugin#1598]
<p>While setting up a new windows build slave on a rented virtual windows server (tried  server2012 as well as server2016) I keep running into this problem where Jenkins jobs cannot check out from git on the new agent:</p>

<p>stderr: error: cannot run c:\msys64\home\jenkins\jenkins\workspace\test-tools\test@​tmp\ssh7559664601978999582.bat: No such file or directory</p>

<p>The required batch file is actually produced and quickly deleted again by the jenkins slave. A script I wrote was able to copy the batch file to a different location before it was deleted. It contains:</p>

<p> </p>

<pre>
@​echo off

<span class="code-quote">"C:\msys64\usr\bin\ssh.exe"</span> -i <span class="code-quote">"C:\msys64\home\jenkins\jenkins\workspace\test-tools\test@​tmp\ssh7913781371770019287.key"</span> -l <span class="code-quote">"git"</span> -o <span class="code-quote">"StrictHostKeyChecking=no"</span> %*
</pre>
<p>The referenced .key file also exists for a short time and contains the correct private key for the git repository.</p>

<p>It is not clear to me if the file is deleted again before it is invoked, or if it is produced only after it was invoked. I would appreciate pointers how to debug this. Can I introduce wait times after file creation? Since this behaviour is dependent on the windows agent, Windows specific filesystem caching and timing may play a role.</p>

<p>I was able to reproduce this with three different virtual windows servers: 2 windows cloud servers rented at ionos.de (2012 and 2016) and a 2016 server rented from a different provider. When executing this in a virtualbox windows 10 machine running on top of ubuntu on my own hardware, this issue does not show up.</p>

<p>Steps to reproduce:</p>
<ul>
	<li>Rent a virtual windows server, e.g. from ionos.de</li>
	<li>Install msys2 and install git and ssh through msys2</li>
	<li>Install openjdk12 from zip, add to PATH and set JAVA_HOME</li>
	<li>Define new build agent NEW-WINDOWS-SERVER in Jenkins</li>
	<li>Start Jenkins slave on windows server like this:
	<ul>
		<li>start msys2 bash shell</li>
		<li>inside bash shell, create ssh tunnel to Jenkins host:
		<ul>
			<li>ssh -L 8080:localhost:8080 -L 35900:localhost:35900 -N -f jenkins@​my.jenkinshost.example.com</li>
		</ul>
		</li>
		<li>Download slave.jar from Jenkins
		<ul>
			<li>wget <a href="http://localhost:8080/jnlpJars/slave.jar" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/jnlpJars/slave.jar</a></li>
		</ul>
		</li>
		<li>Start Jenkins slave
		<ul>
			<li>java -jar slave.jar -jnlpUrl <a href="http://localhost:8080/computer/NEW-WINDOWS-SERVER/slave-agent.jnlp" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/computer/NEW-WINDOWS-SERVER/slave-agent.jnlp</a></li>
		</ul>
		</li>
	</ul>
	</li>
	<li>Create test job in Jenkins, which checks out a git repository from a git repository over ssh (somethingj like ssh://git@​example.com/example-repo)</li>
	<li>The build steps do not matter, add something simple</li>
	<li>Build now</li>
</ul>

<p>attaching the complete console output. Jenkins version: 2.177. Git client plugin version: 2.7.7</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=t_herzke_2017">t_herzke_2017</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-57542" target="_blank">git checkout over ssh fails on windows agents started from msys2 - missing bat file</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: git-client-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2019-05-20T19:46:36+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 20251211-071809
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>While setting up a new windows build slave on a rented virtual windows server (tried  server2012 as well as server2016) I keep running into this problem where Jenkins jobs cannot check out from git on the new agent:</p>

<p>stderr: error: cannot run c:\msys64\home\jenkins\jenkins\workspace\test-tools\test@tmp\ssh7559664601978999582.bat: No such file or directory</p>

<p>The required batch file is actually produced and quickly deleted again by the jenkins slave. A script I wrote was able to copy the batch file to a different location before it was deleted. It contains:</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">@echo off

<span class="code-quote">"C:\msys64\usr\bin\ssh.exe"</span> -i <span class="code-quote">"C:\msys64\home\jenkins\jenkins\workspace\test-tools\test@tmp\ssh7913781371770019287.key"</span> -l <span class="code-quote">"git"</span> -o <span class="code-quote">"StrictHostKeyChecking=no"</span> %*
</pre>
</div></div>
<p>The referenced .key file also exists for a short time and contains the correct private key for the git repository.</p>

<p>It is not clear to me if the file is deleted again before it is invoked, or if it is produced only after it was invoked. I would appreciate pointers how to debug this. Can I introduce wait times after file creation? Since this behaviour is dependent on the windows agent, Windows specific filesystem caching and timing may play a role.</p>

<p>I was able to reproduce this with three different virtual windows servers: 2 windows cloud servers rented at ionos.de (2012 and 2016) and a 2016 server rented from a different provider. When executing this in a virtualbox windows 10 machine running on top of ubuntu on my own hardware, this issue does not show up.</p>

<p>Steps to reproduce:</p>
<ul>
	<li>Rent a virtual windows server, e.g. from ionos.de</li>
	<li>Install msys2 and install git and ssh through msys2</li>
	<li>Install openjdk12 from zip, add to PATH and set JAVA_HOME</li>
	<li>Define new build agent NEW-WINDOWS-SERVER in Jenkins</li>
	<li>Start Jenkins slave on windows server like this:
	<ul>
		<li>start msys2 bash shell</li>
		<li>inside bash shell, create ssh tunnel to Jenkins host:
		<ul>
			<li>ssh -L 8080:localhost:8080 -L 35900:localhost:35900 -N -f jenkins@my.jenkinshost.example.com</li>
		</ul>
		</li>
		<li>Download slave.jar from Jenkins
		<ul>
			<li>wget <a href="http://localhost:8080/jnlpJars/slave.jar" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/jnlpJars/slave.jar</a></li>
		</ul>
		</li>
		<li>Start Jenkins slave
		<ul>
			<li>java -jar slave.jar -jnlpUrl <a href="http://localhost:8080/computer/NEW-WINDOWS-SERVER/slave-agent.jnlp" class="external-link" target="_blank" rel="nofollow noopener">http://localhost:8080/computer/NEW-WINDOWS-SERVER/slave-agent.jnlp</a></li>
		</ul>
		</li>
	</ul>
	</li>
	<li>Create test job in Jenkins, which checks out a git repository from a git repository over ssh (somethingj like ssh://git@example.com/example-repo)</li>
	<li>The build steps do not matter, add something simple</li>
	<li>Build now</li>
</ul>

<p>attaching the complete console output. Jenkins version: 2.177. Git client plugin version: 2.7.7</p></pre>
</details>
</details>
<details><summary><i>1 attachment</i></summary>

- [console.log](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/47414/console.log)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-57542?focusedId=367257&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367257">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>You might be able to compile a custom version of the git client plugin which includes more diagnostic messages in the batch file with steps like:</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>IF EXISTS "C:\msys64\home\jenkins\jenkins\workspace\test-tools\test@tmp\ssh7913781371770019287.key" ECHO Found key file
</pre>
</div></div>

<p>in case something is causing those files to not be detected.</p>

<p>You might also try google searches for specific cases where programs related to git report "<tt>No such file or directory</tt>" on Windows.  That message seems to indicate in at least <a href="https://github.com/fabiospampinato/vscode-open-in-terminal/issues/2" class="external-link" target="_blank" rel="nofollow noopener">one case</a> that a directory name was expected to be provided in a different format.  Another hint indicated that message might be coming from <tt>bash</tt> rather than from the Windows command processor.  If so, then there may be something incorrect in the git client plugin's detection of which operating system command (<tt>sh</tt> or <tt>bat</tt>) should be used to launch the process.</p>
</pre>
</details>

</ul>
</details>

<p>You might be able to compile a custom version of the git client plugin which includes more diagnostic messages in the batch file with steps like:</p>



```
IF EXISTS "C:\msys64\home\jenkins\jenkins\workspace\test-tools\test@​tmp\ssh7913781371770019287.key" ECHO Found key file

```

<p>in case something is causing those files to not be detected.</p>

<p>You might also try google searches for specific cases where programs related to git report "<tt>No such file or directory</tt>" on Windows.  That message seems to indicate in at least <a href="https://github.com/fabiospampinato/vscode-open-in-terminal/issues/2" class="external-link" target="_blank" rel="nofollow noopener">one case</a> that a directory name was expected to be provided in a different format.  Another hint indicated that message might be coming from <tt>bash</tt> rather than from the Windows command processor.  If so, then there may be something incorrect in the git client plugin's detection of which operating system command (<tt>sh</tt> or <tt>bat</tt>) should be used to launch the process.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>t_herzke_2017:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-57542?focusedId=367261&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367261">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for the suggestions. They are reasonable and I would follow up on them but I think I have found a workaround since reporting this issue:</p>

<p>I have now observed that if I invoke the "java -jar slave.jar ..."  on the windows server agent not from the msys2 bash shell but from a git-bash shell, then this problem does not occur and the git checkout over ssh succeeds.</p>

<p>I have now different problems because the actual Jenkins jobs require the invocation of tools from the msys2 environment which are difficult to invoke from the git-bash environment, but I think I will be able to solve this.</p>

<p>The relevant difference for this issue that is introduced by the workaround is probably the different set of environment variables passed on to the java slave process by the different bash shells.</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Thanks for the suggestions. They are reasonable and I would follow up on them but I think I have found a workaround since reporting this issue:</p>

<p>I have now observed that if I invoke the "java -jar slave.jar ..."  on the windows server agent not from the msys2 bash shell but from a git-bash shell, then this problem does not occur and the git checkout over ssh succeeds.</p>

<p>I have now different problems because the actual Jenkins jobs require the invocation of tools from the msys2 environment which are difficult to invoke from the git-bash environment, but I think I will be able to solve this.</p>

<p>The relevant difference for this issue that is introduced by the workaround is probably the different set of environment variables passed on to the java slave process by the different bash shells.</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-57542?focusedId=367272&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367272">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The Windows agents where I test and develop the git plugin are run from either:</p>
<ul>
	<li>Windows batch file launched from a command processor window (a "JNLP agent")</li>
	<li>Windows command processor launched through the OpenSSH server provided by Windows (an ssh agent)</li>
</ul>


<p>I prefer the ssh agent because it can be launched from the Jenkins server.</p>

<p>I've never attempted to run the agent from a git bash shell on Windows.  I have never heard of anyone else using that technique either.  Is there a specific reason you chose to start the agent from a git bash shell instead of using the Windows command processor?</p>
</pre>
</details>

</ul>
</details>

<p>The Windows agents where I test and develop the git plugin are run from either:</p>
<ul>
	<li>Windows batch file launched from a command processor window (a "JNLP agent")</li>
	<li>Windows command processor launched through the OpenSSH server provided by Windows (an ssh agent)</li>
</ul>


<p>I prefer the ssh agent because it can be launched from the Jenkins server.</p>

<p>I've never attempted to run the agent from a git bash shell on Windows.  I have never heard of anyone else using that technique either.  Is there a specific reason you chose to start the agent from a git bash shell instead of using the Windows command processor?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>t_herzke_2017:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-57542?focusedId=367297&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367297">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I can see how running jenkins agent from a unix shell on windows is unusual. </p>

<p>Reason is my builds are cross-platform (linux, mac, windows) but require bash on every platform. I have found in the past that starting the jenkins agent as a child process of the shell I want to use gives me the least trouble to achieve this on windows. I have used, cygwin, msys, msys2, to achieve this so far with no trouble.</p>

<p>Starting Jenkins agent from batch file is an alternative that I'm investigating now but the shell commands from Jenkinsfiles like sh "make sometarget" and their child processes show strange behaviour when I do this, like pretending they were called without parameters.</p>
</pre>
</details>

</ul>
</details>

<p>I can see how running jenkins agent from a unix shell on windows is unusual. </p>

<p>Reason is my builds are cross-platform (linux, mac, windows) but require bash on every platform. I have found in the past that starting the jenkins agent as a child process of the shell I want to use gives me the least trouble to achieve this on windows. I have used, cygwin, msys, msys2, to achieve this so far with no trouble.</p>

<p>Starting Jenkins agent from batch file is an alternative that I'm investigating now but the shell commands from Jenkinsfiles like sh "make sometarget" and their child processes show strange behaviour when I do this, like pretending they were called without parameters.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>t_herzke_2017:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-57542?focusedId=367349&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367349">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Workaround recipe:</p>

<p>1) Identify additional environment variables and their values when starting the msys2 bash via start menu vs by invoking bash.exe from cmd.exe.</p>

<p>2) Create bash skript that sets these environment variables in multiple <tt>export NAME=value</tt> lines, also add the extended PATH value. Store as <tt>jenkins.env</tt> in msys2 root folder.</p>

<p>3) In the Jenkinsfiles, define custom step windows_bash</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def windows_bash(command) {
  bat (<span class="code-quote">'C:\\msys64\\usr\\bin\\bash -c <span class="code-quote">"source /jenkins.env &amp;&amp; '</span> + command + <span class="code-quote">' "</span>'</span>)
}
</pre>
</div></div>
<p>4) In the Jenkinsfiles, define a delegating custom step that forwards to either sh or windows_bash depending on the operating system of the Agent</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def bash(command) {
  windows() ? windows_bash(command) : sh(command)
}
</pre>
</div></div>
<p>where windows() has to be defined to return true only for windows agents.</p>

<p>5) In all places in the Jenkinsfile where a shell command could be executed either on windows or on Unix, use <tt>bash "command"</tt> instead of <tt>sh "command"</tt></p>
</pre>
</details>

</ul>
</details>

<p>Workaround recipe:</p>

<p>1) Identify additional environment variables and their values when starting the msys2 bash via start menu vs by invoking bash.exe from cmd.exe.</p>

<p>2) Create bash skript that sets these environment variables in multiple <tt>export NAME=value</tt> lines, also add the extended PATH value. Store as <tt>jenkins.env</tt> in msys2 root folder.</p>

<p>3) In the Jenkinsfiles, define custom step windows_bash</p>

<pre>
def windows_bash(command) {
  bat (<span class="code-quote">'C:\\msys64\\usr\\bin\\bash -c <span class="code-quote">"source /jenkins.env && '</span> + command + <span class="code-quote">' "</span>'</span>)
}
</pre>
<p>4) In the Jenkinsfiles, define a delegating custom step that forwards to either sh or windows_bash depending on the operating system of the Agent</p>

<pre>
def bash(command) {
  windows() ? windows_bash(command) : sh(command)
}
</pre>
<p>where windows() has to be defined to return true only for windows agents.</p>

<p>5) In all places in the Jenkinsfile where a shell command could be executed either on windows or on Unix, use <tt>bash "command"</tt> instead of <tt>sh "command"</tt></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>t_herzke_2017:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-57542?focusedId=367350&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-367350">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The problem occurs under special circumstances that are probably outside of the scope of Jenkins. Closed as won't-fix.</p>
</pre>
</details>

</ul>
</details>

<p>The problem occurs under special circumstances that are probably outside of the scope of Jenkins. Closed as won't-fix.</p>
